source/util/move_to_front.cpp

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "move_to_front.h"
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+
+namespace spvutils {
+
+size_t MoveToFront::RankFromId(uint32_t id) {
+  const size_t old_size = GetSize();
+  (void)old_size;
+  const auto it = id_to_node_.find(id);
+
+  if (it == id_to_node_.end()) {
+    assert(id == next_id_);
```

Comment:
I think this assertion is weird.  It's a strange coupling between this MTF sequence and the client.

I think you've written that the expected operation is for the client to call IdForRank(0) to generate new Ids.   But since you have a simple id allocation scheme, this forces RankFromId to be restrictive in this way.